### PR TITLE
JDK-8283807: Handle CompileThreshold the same as other thresholds when scaled with -XX:CompileThresholdScaling

### PR DESCRIPTION
**Pull request by tobiasholenstein**
`CompileThresholdScaling` scales several compiler thresholds. If the scaled threshold value is outside of the allowed range, a warning is printed and the default value is kept. But when scaling `CompileThreshold` we do not check the scaled value because it is not a ranged flag. Later when verifying flags with a constraint function `OnStackReplacePercentageConstraintFunc` which calls `CompileThresholdConstraintFunc` fails because `CompileThreshold` is outside of the allowed range. This stops the VM from being created. 

The fix is to call `CompileThresholdConstraintFunc` when scaling `CompileThreshold` with `CompileThresholdScaling`. If `CompileThresholdConstraintFunc` fails we print a warning and keep the default value of `CompileThreshold`. This behaviour is then consistent with the other 15 other thresholds where the VM just prints a warning if the scaled value is too large and then keeps the current value.

We also have to exclude `CompileThresholdScaling` from max. range testing in `TestOptionsWithRanges.java`, because
it is expected to print "outside the allowed range" warnings for the scaled flag and the "outside the allowed range" warning does not refer to `CompileThresholdScaling` itself. 

One question that remains it why `CompileThresholdScaling` takes the default value when a scaled threshold is too large instead of the max. value?  Because now we can have the case where some compiler thresholds are scaled to almost their max. value and others that are already larger than the maximum are default again (which can be the minimum value in some cases). I suggest to address this issue in a separate RFE.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8283807](https://bugs.openjdk.java.net/browse/JDK-8283807): Handle CompileThreshold the same as other thresholds when scaled with -XX:CompileThresholdScaling

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [834cfcfb](https://git.openjdk.java.net/jdk/pull/8501/files/834cfcfbebcf07b0338e61694e1f68b861297a68)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8501/head:pull/8501` \
`$ git checkout pull/8501`

Update a local copy of the PR: \
`$ git checkout pull/8501` \
`$ git pull https://git.openjdk.java.net/jdk pull/8501/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8501`

View PR using the GUI difftool: \
`$ git pr show -t 8501`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8501.diff">https://git.openjdk.java.net/jdk/pull/8501.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good.

File: test/hotspot/jtreg/runtime/CommandLine/OptionsValidation/TestOptionsWithRanges.java [hunk 1]
```diff
@@ -257,6 +257,13 @@ public static void main(String[] args) throws Exception {
         excludeTestMaxRange("NonNMethodCodeHeapSize");
         excludeTestMaxRange("CodeCacheExpansionSize");
 
+        /*
+         * Exclude CompileThresholdScaling from max range testing, because
+         * it is expected to print "outside the allowed range" warnings for
+         * the scaled flags
+         */
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/CommandLine/OptionsValidation/TestOptionsWithRanges.java:264]: This comment is not very clear. Do you mean

```
Max range tests expect a warning message of "outside the allowed range"
but CompileThresholdScaling prints a different warning message when it's
set too high.
```

**Review by iklam — CHANGES_REQUESTED** [review]
Could you add a test case with several large values for -XX:CompileThresholdScaling and check that the warning message is printed?

File: test/hotspot/jtreg/runtime/CommandLine/OptionsValidation/TestOptionsWithRanges.java (hunk 1 above)
Comment (tobiasholenstein) [reply test/hotspot/jtreg/runtime/CommandLine/OptionsValidation/TestOptionsWithRanges.java:265]: I changed to comment. I hope it is clearer now

**Comment by tobiasholenstein** [comment]
> Could you add a test case with several large values for -XX:CompileThresholdScaling and check that the warning message is printed?

I added a test now

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Review by iklam — APPROVED** [review]
LGTM

**Comment by tobiasholenstein** [comment]
@vnkozlov , @TobiHartmann and @iklam thanks for the reviews!

**Pull request closed without merging**